src/adc.rs

```diff
@@ -0,0 +1,680 @@
+use crate::hal::adc::{Channel, OneShot};
+use crate::hal::blocking::delay::DelayUs;
+
+use crate::stm32::{ADC1, ADC2, ADC3};
+
+use crate::gpio::Analog;
+use crate::gpio::gpioa::{PA0, PA1, PA2, PA3, PA4, PA5, PA6, PA7};
+use crate::gpio::gpiob::{PB0, PB1};
+use crate::gpio::gpioc::{PC0, PC1, PC2, PC3, PC4, PC5};
+use crate::gpio::gpiof::{PF3, PF4, PF5, PF6, PF7, PF8, PF9, PF10, PF11, PF12, PF13, PF14};
+use crate::gpio::gpioh::{PH2, PH3, PH4, PH5};
+use crate::delay::Delay;
+use crate::rcc::Ccdr;
+use crate::rcc::{AHB1, AHB4, D3CCIPR};
+
+pub struct Adc<ADC> {
+    rb: ADC,
+    sample_time: AdcSampleTime,
+    resolution: AdcSampleResolution,
+    lshift: AdcLshift,
+}
+
+/// ADC sampling time
+///
+/// Options for the sampling time, each is T + 0.5 ADC clock cycles.
+//
+// Refer to RM0433 Rev 6 - Chapter 24.4.13
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[allow(non_camel_case_types)]
+pub enum AdcSampleTime {
+    /// 1.5 cycles sampling time
+    T_1,
+    /// 2.5 cycles sampling time
+    T_2,
+    /// 8.5 cycles sampling time
+    T_8,
+    /// 16.5 cycles sampling time
+    T_16,
+    /// 32.5 cycles sampling time
+    T_32,
+    /// 64.5 cycles sampling time
+    T_64,
+    /// 387.5 cycles sampling time
+    T_387,
+    /// 810.5 cycles sampling time
+    T_810,
+}
+
+impl AdcSampleTime {
+    pub fn default() -> Self {
+        AdcSampleTime::T_32
+    }
+}
+
+// Refer to RM0433 Rev 6 - Chapter 24.4.13
+impl From<AdcSampleTime> for u8 {
+    fn from(val: AdcSampleTime) -> u8 {
+        match val {
+            AdcSampleTime::T_1 => 0b000,
+            AdcSampleTime::T_2 => 0b001,
+            AdcSampleTime::T_8 => 0b010,
+            AdcSampleTime::T_16 => 0b011,
+            AdcSampleTime::T_32 => 0b100,
+            AdcSampleTime::T_64 => 0b101,
+            AdcSampleTime::T_387 => 0b110,
+            AdcSampleTime::T_810 => 0b111,
+        }
+    }
+}
+
+/// ADC sampling resolution
+///
+/// Options for sampling resolution
+//
+// Refer to RM0433 Rev 6 - Chapter 24.2
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[allow(non_camel_case_types)]
+pub enum AdcSampleResolution {
+    /// 16 bit resulution
+    B_16,
```

Comment:
Here would be a good place to use the stm32-rs enum definition. You can do this like:

```rust
/// Resolution to sample at
pub type Resolution = crate::stm32::adc3::cfgr::RESR;
```

The definition is then the automatically generated one from [here.](https://github.com/stm32-rs/stm32-rs/blob/5142af62d53343c8110eaac1ff55d167c32b07c0/peripherals/adc/adc_h7_revision_y.yaml#L5)

This will save the enum definition and the From for u8 implementation. The default will then be `Resolution::SIXTEENBIT`.

Reply:
I'd usually agree with you on this, however this specific piece of code is revision dependent as the config for the 8 bit resolution is dependent on the revision (as can be seen below), however this code is generated for revision y and would thus not be compatible with rev v

Reply:
The intention is that when stm32-rs does add support for Rev V, this crate will pass a `rev_v` feature gate back to stm32-rs. In that case, stm32-rs will have the correct definition.

This means that the difference between revisions only needs to be written and maintained once instead of twice.

Tagging tracking issue #9

Reply:
Okay so I tried to implement this approach, however in the example codes this causes some problems:
```
error: enum variants on type aliases are experimental
  --> examples/adc.rs:49:25
   |
49 |     adc1.set_resolution(adc::Resolution::SIXTEENBIT);
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(type_alias_enum_variants)]` to the crate attributes to enable

error: aborting due to previous error
```

Reply:
Add `#![feature(type_alias_enum_variants)]` to each example as well. The lib has it already.